pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,237 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or impliec.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ngrok
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ngrok/ngrok-api-go/v7"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/ptr"
+
+	esv1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1"
+)
+
+const (
+	defaultDescription = "Managed by External Secrets Operator"
+)
+
+var (
+	errWriteOnlyOperations     = errors.New("not implemented - the ngrok provider supports write-only operations")
+	errVaultDoesNotExist       = errors.New("vault does not exist")
+	errVaultSecretDoesNotExist = errors.New("vault secret does not exist")
+	errCannotPushNilSecret     = errors.New("cannot push nil secret")
+)
+
+type VaultClient interface {
+	Create(context.Context, *ngrok.VaultCreate) (*ngrok.Vault, error)
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Vault]
+}
+
+type SecretsClient interface {
+	Create(context.Context, *ngrok.SecretCreate) (*ngrok.Secret, error)
+	Update(context.Context, *ngrok.SecretUpdate) (*ngrok.Secret, error)
+	Delete(context.Context, string) error
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Secret]
+}
+
+type client struct {
+	vaultClient   VaultClient
+	secretsClient SecretsClient
+	vaultName     string
+}
+
+func (c *client) PushSecret(ctx context.Context, secret *corev1.Secret, data esv1.PushSecretData) error {
+	if secret == nil {
+		return errCannotPushNilSecret
+	}
+
+	// First, get the vault by name. If it doesn't exist, create it.
+	vault, err := c.getVaultByName(ctx, c.vaultName)
+	if err != nil {
+		if !errors.Is(err, errVaultDoesNotExist) {
+			return err
+		}
+
+		// Create the vault if it does not exist
+		vault, err = c.vaultClient.Create(ctx, &ngrok.VaultCreate{
+			Name:        c.vaultName,
+			Description: defaultDescription,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create vault: %w", err)
+		}
+	}
+
+	// Prepare the secret data for pushing
+	value, err := json.Marshal(secret.Data)
+	if err != nil {
+		return fmt.Errorf("json.Marshal failed with error: %w", err)
+	}
+
+	// If key is specified, get the value from the secret data
+	if data.GetSecretKey() != "" {
+		var ok bool
+		value, ok = secret.Data[data.GetSecretKey()]
+		if !ok {
+			return fmt.Errorf("key %s not found in secret", data.GetSecretKey())
+		}
+	}
```

Comment:
I don't think it is nice to both marshal secret.Data and then to limit secret.Data to only one secret key if given. I think you should move the marshalling of the whole secret to the else block

Reply:
Thanks for the catch there. Fixed in https://github.com/external-secrets/external-secrets/pull/5160/commits/cc27c2bc90dfa0d33be2d2fd0fba367d224d572b